File: backend/app/services/overpass.py
"""OpenStreetMap via Overpass — live per query (spec §4). We ask for a
curated slice of community-tagged oddities (tourism/historic tags) inside
bboxes tiled along the route corridor, never one huge box."""

import logging

import httpx

from .. import config, scoring
from ..categories import categorize
from ..models import Place

log = logging.getLogger(__name__)

# Tag values that match the "quirky deep cuts" brief; deliberately excludes
# generic tourism (hotels, viewpoints, info boards) and ubiquitous
# historic=memorial.
NODE_FILTERS = [
    '["tourism"~"^(attraction|artwork)$"]["name"]',
    '["historic"~"^(ruins|fort|wreck|ship|aircraft|locomotive|mine|battlefield|archaeological_site)$"]["name"]',
]


def _build_query(bboxes: list[tuple[float, float, float, float]]) -> str:
    parts = []
    for (min_lng, min_lat, max_lng, max_lat) in bboxes:
        bbox = f"({min_lat:.4f},{min_lng:.4f},{max_lat:.4f},{max_lng:.4f})"
        for flt in NODE_FILTERS:
            parts.append(f"node{flt}{bbox};")
            parts.append(f"way{flt}{bbox};")
    body = "\n".join(parts)
    return f"[out:json][timeout:50];\n(\n{body}\n);\nout center tags 900;"


def _element_to_place(el: dict) -> Place | None:
    tags = el.get("tags", {})
    name = tags.get("name")
    if not name:
        return None
    if el["type"] == "node":
        lat, lng = el.get("lat"), el.get("lon")
    else:
        center = el.get("center") or {}
        lat, lng = center.get("lat"), center.get("lon")
    if lat is None or lng is None:
        return None
    has_wiki = "wikipedia" in tags or "wikidata" in tags
    desc = tags.get("description", "") or tags.get("inscription", "")[:200]
    image = tags.get("image") if (tags.get("image", "").startswith("http")) else None
    tag_list = [
        f"{k}={v}" for k, v in tags.items()
        if k in ("tourism", "historic", "artwork_type", "ruins", "man_made")
    ]
    score = scoring.universal_boosts(
        scoring.osm(len(tags), has_wiki), has_image=bool(image), description=desc
    )
    url = (
        f"https://en.wikipedia.org/wiki/{tags['wikipedia'].split(':', 1)[-1]}"
        if tags.get("wikipedia") and ":" in tags.get("wikipedia", "")
        else f"https://www.openstreetmap.org/{el['type']}/{el['id']}"
    )
    return Place(
        id=f"osm:{el['type']}/{el['id']}",
        name=name,
        description=desc,
        category=categorize(tag_list, name, desc),
        source="osm",
        source_url=url,
        lat=lat,
        lng=lng,
        image_url=image,
        score=score,
        tags=tag_list,
    )


async def fetch(bboxes: list[tuple[float, float, float, float]]) -> list[Place]:
    query = _build_query(bboxes)
    try:
        async with httpx.AsyncClient(timeout=60) as client:
            r = await client.post(
                config.OVERPASS_URL,
                data={"data": query},
                headers={"User-Agent": config.USER_AGENT},
            )
            r.raise_for_status()
            elements = r.json().get("elements", [])
    except Exception as e:
        log.warning("Overpass query failed (continuing without OSM): %s", e)
        return []
    places = []
    seen = set()
    for el in elements:
        p = _element_to_place(el)
        if p and p.id not in seen:
            seen.add(p.id)
            places.append(p)
    return places

File: worker/services/overpass.ts
// OpenStreetMap via Overpass — live per query (spec §4). Mirrors
// backend/app/services/overpass.py: a curated slice of community-tagged
// oddities (tourism/historic tags) inside bboxes tiled along the route
// corridor, never one huge box.

import { categorize } from "../categories";
import { DEFAULTS, userAgent } from "../config";
import * as scoring from "../scoring";
import type { Env, Place } from "../types";

// Tag values that match the "quirky deep cuts" brief; deliberately excludes
// generic tourism (hotels, viewpoints, info boards) and ubiquitous
// historic=memorial.
const NODE_FILTERS = [
  '["tourism"~"^(attraction|artwork)$"]["name"]',
  '["historic"~"^(ruins|fort|wreck|ship|aircraft|locomotive|mine|battlefield|archaeological_site)$"]["name"]',
];

interface OverpassElement {
  type: string;
  id: number;
  lat?: number;
  lon?: number;
  center?: { lat?: number; lon?: number };
  tags?: Record<string, string>;
}

function buildQuery(bboxes: [number, number, number, number][]): string {
  const parts: string[] = [];
  for (const [minLng, minLat, maxLng, maxLat] of bboxes) {
    const bbox = `(${minLat.toFixed(4)},${minLng.toFixed(4)},${maxLat.toFixed(4)},${maxLng.toFixed(4)})`;
    for (const flt of NODE_FILTERS) {
      parts.push(`node${flt}${bbox};`);
      parts.push(`way${flt}${bbox};`);
    }
  }
  return `[out:json][timeout:50];\n(\n${parts.join("\n")}\n);\nout center tags 900;`;
}

function elementToPlace(el: OverpassElement): Place | null {
  const tags = el.tags ?? {};
  const name = tags["name"];
  if (!name) return null;
  const lat = el.type === "node" ? el.lat : el.center?.lat;
  const lng = el.type === "node" ? el.lon : el.center?.lon;
  if (lat == null || lng == null) return null;
  const hasWiki = "wikipedia" in tags || "wikidata" in tags;
  const desc = tags["description"] || (tags["inscription"] || "").slice(0, 200);
  const image = (tags["image"] || "").startsWith("http") ? tags["image"] : null;
  const tagList = Object.entries(tags)
    .filter(([k]) => ["tourism", "historic", "artwork_type", "ruins", "man_made"].includes(k))
    .map(([k, v]) => `${k}=${v}`);
  const score = scoring.universalBoosts(
    scoring.osm(Object.keys(tags).length, hasWiki), Boolean(image), desc
  );
  const wp = tags["wikipedia"];
  const url = wp && wp.includes(":")
    ? `https://en.wikipedia.org/wiki/${wp.split(/:(.*)/)[1]}`
    : `https://www.openstreetmap.org/${el.type}/${el.id}`;
  return {
    id: `osm:${el.type}/${el.id}`,
    name,
    description: desc,
    category: categorize(tagList, name, desc),
    source: "osm",
    sources: [],
    source_url: url,
    lat,
    lng,
    image_url: image,
    score,
    tags: tagList,
  };
}

export async function fetchPlaces(
  env: Env, bboxes: [number, number, number, number][]
): Promise<Place[]> {
  const query = buildQuery(bboxes);
  let elements: OverpassElement[];
  try {
    const r = await fetch(env.OVERPASS_URL || DEFAULTS.OVERPASS_URL, {
      method: "POST",
      body: new URLSearchParams({ data: query }),
      headers: { "User-Agent": userAgent(env) },
    });
    if (!r.ok) throw new Error(`status ${r.status}`);
    elements = ((await r.json()) as { elements?: OverpassElement[] }).elements ?? [];
  } catch (e) {
    console.warn(`Overpass query failed (continuing without OSM): ${e}`);
    return [];
  }
  const places: Place[] = [];
  const seen = new Set<string>();
  for (const el of elements) {
    const p = elementToPlace(el);
    if (p && !seen.has(p.id)) {
      seen.add(p.id);
      places.push(p);
    }
  }
  return places;
}
